core/blockchain.go

```diff
@@ -1736,3 +1743,47 @@ func (bc *BlockChain) SubscribeLogsEvent(ch chan<- []*types.Log) event.Subscript
 func (bc *BlockChain) SubscribeBlockProcessingEvent(ch chan<- bool) event.Subscription {
 	return bc.scope.Track(bc.blockProcFeed.Subscribe(ch))
 }
+
+// GetAncestorHash return the hash of ancestor at the given number
+func (bc *BlockChain) GetAncestorHash(ref *types.Header, target uint64) common.Hash {
+	number := ref.Number.Uint64() - 1
+	hash := ref.ParentHash
+	if target == number {
+		return hash
+	}
+	if target > number {
+		// Should never happen
+		log.Error("Ancestor number must be <= descendant", "ancestor target", target, "descendant", number)
+		return common.Hash{}
+	}
+	// Iterate the chain until we hit the target or we hit a ancestor
+	// within the storage
+	for !bc.hashHistory.Is(number, hash) {
+		header := bc.GetHeader(hash, number)
+		if header == nil {
+			return common.Hash{}
+		}
+		number = header.Number.Uint64() - 1
+		hash = header.ParentHash
+		if number == target {
+			return hash
+		}
+	}
+	// The hash storage has the right ancestor chain
+	if h, found := bc.hashHistory.Get(target); found {
+		return h
+	}
+	// The number is too old - continue iteration from the
+	// earliest point
+	number, hash = bc.hashHistory.Oldest()
+	// At this point we iterate the remaining headers via the chain
+	for number > target {
+		header := bc.GetHeader(hash, number)
+		if header == nil {
+			return common.Hash{}
+		}
+		number = header.Number.Uint64() - 1
+		hash = header.ParentHash
+	}
+	return hash
```

Comment:
Is it true that there are cases where `hash` will not correspond to `target`?

Reply:
Yes. If we get an old sidechain. If hashhistory contains e..g blocks [`1000` .. `1288`], and we import sideblock on `1100` which requests block `900`. Then the hash history can be used to jump back to block `1000`, but the remaining `100` headers are iterated by this loop

Reply:
Right. Because to get here we have already established we are on the canonical chain. Thanks!

Reply:
Since we've established we are on the canonical chain, doesn't this mean that the `for` loop can be replaced with `rawdb.ReadCanonicalHash` (or a wrapper around it)?

Reply:
It could, but it creates the invariant that the hash storage is used to track canon blocks. Right now, there's no such invariant _within the hash storage itself_ 

I'll think about it

Reply:
Canonical blocks will be the common case, so its something to consider.